### ISSUE TITLE
Honor zigbee2mqtt device availability

### ISSUE THREAD

**Issue by economycomfort** [itavero/homebridge-z2m#36]
**Problem:**
Some Zigbee devices are at risk of losing power at the mains, and not by telling the device to turn on or off.  Think of a bulb in a light fixture controlled by a wall switch.

If I turn that wall switch off, there's no outward indication that the Zigbee device is having an issue, it just never responds to commands.

Since most MQTT commands are one-way, I can send all the commands to the device I want through Homekit (and thus through Homebridge -> homebridge-z2m), and the status in Homekit will happily change to whatever I tell it... however, nothing happens at the device, of course, because there's no two way communication from the Zigbee device.

**Suggested solution:**
There's a feature in zigbee2mqtt called "availability" (https://www.zigbee2mqtt.io/information/availability.html).  When enabled, it will "ping" a Zigbee device to determine if it's responding to commands or not.  If the device doesn't respond, it can mark the device properly as "offline" in Homekit until one of those pings succeeds (or a command is sent again).

**Describe alternatives you've considered**
Not really sure this applies -- I think it's a limitation of homebridge-z2m currently not looking for it.  I'll admit I'm not an expert in MQTT plumbing, but I suspect this to be the case.

From zigbee2mqtt logs:

`debug 2021-01-07 18:00:55: Received Zigbee message from 'Sengled Test Bulb', type 'readResponse', cluster 'genBasic', data '{"zclVersion":1}' from endpoint 1 with groupID null`
`debug 2021-01-07 18:00:55: Successfully pinged 'Sengled Test Bulb'`

To this end, I would assume there's a way for homebridge-z2m to be able to pick up on these messages as well, and impact the state of the end device.

**Additional context**
n/a

**Are you willing to contribute to this feature?**
If I can.  I'm no developer but I can more or less tell what code is doing.  Though I don't have a lot of experience in Javascript (mostly python).

**Comment by zerobaud** [comment]
I'm actually going to keep messing with this before I say it doesn't work, but I've not yet seen it proliferate all the way through to homekit.  Since I know that the functionality is supported through the zigbee2mqtt layer, the next unknown is homebridge-z2m, which is why I'm posting it here.

In Homebridge logs, I never see anything show up other than publish requests to a given topic, so I'm making an assumption that it's not actively polling device state.

**Comment by itavero** [comment]
This is not in plugin at the moment. First thing on my list in this area is looking at the overall zigbee2mqtt bridge state.
After that I'll have a mechanism to let the service handlers know if an accessory is online or not, which can probably also be used for this.

The "annoying" thing is that, as far as I understood, there's no way in Homebridge for a plugin to indicate this accessory is offline. It seems that the only thing you can do is return an error when characteristics are get/set, so it's a bit tedious.

**Comment by zerobaud** [comment]
Gotcha, thanks for the clarification!

From a place of ignorance, would a workaround here be for homebridge-z2m to observe the ping state from zigbee2mqtt for a given device, and if that ping state is 'offline', then basically publish a state of "off" to that same device?  It would then show as "Offline" in Homekit, which is actually true, and the message could be published roughly at the same time frame specified in the availability_timeout setting for zigbee2mqtt.

Of course, the best way would be for Homekit to display "No Response" or something similar, but if there's no way to duplicate that, I'd think showing "Offline" would be a close runner-up.

This way, you wouldn't necessarily see an error, but you'd see that the device keeps going offline in Homekit when it's not connected to power or is otherwise unavailable.

Let me know if that makes sense?

**Comment by itavero** [comment]
It is possible to reproduce the _No Response_ behavior, I believe.
It just requires some more code and thought 😅

**Comment by itavero** [comment]
I've started working on the implementation today.
It seems I have a working solution, however I still need to fix/update the automated tests.

I'm also still thinking about an easier way to provide this functionality. Currently you have to think about this in every `ServiceHandler` implementation (a.k.a. whenever you configure a HomeKit characteristic). I'll see if I can come up with a way that is more generic. Perhaps just iterating over all services and characteristics _after_ all the `ServiceHandler`s have been created.

**Comment by zerobaud** [comment]
I appreciate it!  I think it would go a long way for usability in general of the plugin.  By far one of the least effort Homebridge plugins I've found to date.  I'll pull the :next branch when its ready for test for sure!

**Comment by itavero** [comment]
Note to self: I should also verify that `availability_timeout != 0` in the zigbee2mqtt configuration (which is also published on `zigbee2mqtt/bridge/info` and maybe also consider the `availability_passlist` and `availability_blocklist` when processing the incoming MQTT messages.

The reason for this is that old MQTT topics/messages may stick around if the feature is disabled in zigbee2mqtt (as I've just noticed).

**Comment by itavero** [comment]
I saw some more questions from other plugin developers on the Homebridge Discord server.
Something tells me that there should probably be an API added in Homebridge to get an easy and unified way for marking devices as _unavailable_ instead of adding additional logic to each characteristic.
I'll think on this a bit more and, once I find some time to do so, dive into the Homebridge sources to write a feature proposal for Homebridge.

**Comment by itavero** [comment]
Just wanted to let you know, this issue is still on my mind, but I haven't found the time yet to propose a supporting feature/API for this to the Homebridge community.

**Comment by itavero** [comment]
Perhaps it is possible to use the [Reachable](https://developers.homebridge.io/#/characteristic/Reachable) characteristic for this. I think it used to be integrated in HAP-NodeJS, but I got removed at some point (not sure why).
It's at least worth the experiment ;)

**Comment by itavero** [comment]
The Bridging State service and Reachable characteristic no longer work with recent versions of iOS unfortunately.

I think I'll need to add a generic way of adding a handler to all characteristics of a service to deal with checking the online/offline state.

**Comment by belkone** [comment]
hi, are there any updates on this topic? ;)

**Comment by itavero** [comment]
No updates from my side unfortunately.

**Comment by ronlimon** [comment]
This would be the missing link in my extensive zigbee setup - any update from your side?

**Comment by itavero** [comment]
No, but if someone where to provide a clean solution in a PR for reporting errors from all characteristics when a device is not available, I'd definitely look at it.

**Comment by itavero** [comment]
I think it may be possible to add something in the `Zigbee2mqttAccessory` class to detect when characteristics are added (perhaps by using the `SERVICE_CONFIGURATION_CHANGE` event from `Service`) and then add a new/overridden `get` callback.
Not sure if that would work if the converter would also want to add a `get` callback, but so far this is not being used as far as I know.

**Comment by itavero** [comment]
Short update / "note to self": after having another look at the homebridge/HAP-NodeJS sources, I think it might be possible to use `updateValue` to make HAP aware of the fact that an accessory is not reachable (see [this interface](https://github.com/homebridge/HAP-NodeJS/blob/63bc58973cb0e8351b9bba73c9ed138d77ce6eef/src/lib/Characteristic.ts#L1420))

However, this probably still needs to be handled by the service handlers/converters, as they know which characteristics are the "main" characteristic for their service.

The service handler should probably call `characteristic.updateValue(new HapStatusError(HAPStatus.SERVICE_COMMUNICATION_FAILURE))` when the device (or Zigbee2MQTT itself) become unreachable and revert this by calling `characteristic.updateValue(new HapStatusError(HAPStatus.SUCCESS))`.
Even though this still sets `this.status` to a non-null value (in `PlatformAccessory`), I think this should work as the `handleGetRequest` function [only checks `this.statusCode`](https://github.com/homebridge/HAP-NodeJS/blob/63bc58973cb0e8351b9bba73c9ed138d77ce6eef/src/lib/Characteristic.ts#L1577-L1579).

**Comment by itavero** [comment]
Short update: tried to implement based on the info mentioned above. Marking them as _unavailable_ a.k.a. _not responding_ works, but  I can't easily make them available again.
Setting the error to `SUCCESS` does not seem to work correctly. I might try to somehow get the old value of the characteristic and set that again to clear the value, but I might also just be missing something else or testing it wrong.

In other words, some more investigation is needed 😉

**Comment by itavero** [comment]
Resending the last value using `sendEventNotification` does seem to work, but I still have to reopen the _Home.app_ for it to show the devices as responding again. I think that might be acceptable though.

**Comment by belkone** [comment]
I think many accessories behave like this, it is possible that, this is Home.app at all :)

**Comment by itavero** [comment]
Just released `v1.10.0` to `homebridge-z2m@next` which includes this feature. Because there were quite some internal changes, I haven't pushed it to `latest` yet. So far it seems to be working fine on my setup, but if you run into any issues, please do let me know.

**Comment by bulivlad** [comment]
hey, not sure if this is the breaking change but it seems to be one of the candidates. Just updated the plugin from `v1.9.2` where everything was looking good to `v1.10.0` and suddenly started to see loads of errors like 
```
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Received devices...
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
[12/15/2022, 3:22:39 PM] [zigbee2mqtt] Update "error" status for characteristics of <accessory_name_here> to 0
```
in the logs and mostly all of the devices controller by this plugin gone into `No response` status. I also tried to add `"ignore_availability": true,` as a default as well as for every accessory but nothing changed :(

**Comment by itavero** [comment]
@bulivlad This message is not an error actually. Maybe the phrasing is a bit weird. It is indeed to trigger the "No response" behavior in the HomeKit. (Should also only be in the debug logs.)

Setting the "error" to 0 actually means "Success" a.k.a. available, but it's weird that `ignore_availability` does not seem to work for you (I believe it then still sets the error to `0`).

A few additional questions though:
Do you have the Availability feature in Zigbee2MQTT turned on or not?
Was Zigbee2MQTT itself online?

Perhaps you can create a bug report and also include a bit more information on your Zigbee2MQTT setup, the configurations you've tried and the answers to these questions?

**Comment by bulivlad** [comment]
@itavero opened one [here](https://github.com/itavero/homebridge-z2m/issues/593). I think I added all the details in there but please let me know if you need any other details on that
